### PR TITLE
[Feature][Vectorized] Support ES scan on vectorized engine

### PR DESCRIPTION
**Pull request by Gabriel39**
## Proposed changes

Support ES scan on vectorized engine

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Gabriel39** [comment]
cc @HappenLee

File: be/src/exec/es_http_scan_node.h [hunk 1]
```diff
@@ -56,6 +56,8 @@ class EsHttpScanNode : public ScanNode {
     // Write debug string of this into out.
     virtual void debug_string(int indentation_level, std::stringstream* out) const override;
 
+    TupleDescriptor* _tuple_desc;
```
Comment (HappenLee) [inline be/src/exec/es_http_scan_node.h:59]: 这里为什么移动了一下位置

Comment (Gabriel39) [reply be/src/exec/es_http_scan_node.h:59]: 子类VESHttpScan要访问，之前是private，子类访问不到，我看MysqlScanNode的_tuple_desc也是放在protected里的，应该没什么安全问题？

**Comment by Gabriel39** [comment]
> @Gabriel39 Thanks, This is great. Can you provide some performance comparison for after using the `Vectorized` technique？

Sure. I will attach the comparison result below in next few days. Btw, vectorized es scan will be slower than row based es scan as introducing additional data copying (row to column) and it seems like inevitable. I think vectorized es scan only benefit from the vectorized engine in more complex query which does not only contain es scan.

File: be/src/vec/exec/ves_http_scan_node.cpp [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+            _queue_writer_cond.notify_all();
+        }
+    }
+
+    if (_eos) {
+        *eos = true;
+        return Status::OK();
+    }
+
+    if (_scan_finished.load()) {
```
Comment (HappenLee) [inline be/src/vec/exec/ves_http_scan_node.cpp:32]: 这个if可以作为一个||条件放在上一个if上吧

File: be/src/vec/exec/ves_http_scan_node.cpp [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+
+Status VEsHttpScanNode::scanner_scan(std::unique_ptr<EsHttpScanner> scanner,
+                                    const std::vector<ExprContext*>& conjunct_ctxs,
+                                    EsScanCounter* counter) {
+    RETURN_IF_ERROR(scanner->open());
+    bool scanner_eof = false;
+    bool init_columns = false;
+
+    const int batch_size = _runtime_state->batch_size();
+    MemPool* tuple_pool = new MemPool(mem_tracker().get());
```
Comment (HappenLee) [inline be/src/vec/exec/ves_http_scan_node.cpp:180]: 这里直接new会内存泄漏的

File: be/src/vec/exec/ves_http_scan_node.cpp [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+    }
+
+    return Status::OK();
+}
+
+Status VEsHttpScanNode::insert_data(const SlotDescriptor* slot_desc,
+                                    vectorized::MutableColumnPtr* column_ptr,
+                                    void* data) {
+    vectorized::MutableColumnPtr* col_ptr = column_ptr;
+    if (slot_desc->is_nullable()) {
```
Comment (HappenLee) [inline be/src/vec/exec/ves_http_scan_node.cpp:98]: 这里是不可能出现null值的吗？

File: be/src/vec/exec/ves_http_scan_node.cpp (hunk 2 above)
Comment (Gabriel39) [reply be/src/vec/exec/ves_http_scan_node.cpp]: done

File: be/src/vec/exec/ves_http_scan_node.cpp (hunk 3 above)
Comment (Gabriel39) [reply be/src/vec/exec/ves_http_scan_node.cpp]: done.

File: be/src/vec/exec/ves_http_scan_node.cpp (hunk 4 above)
Comment (HappenLee) [reply be/src/vec/exec/ves_http_scan_node.cpp:93]: 你这里是nullable的值，默认都是给nullable列插入0，不是肯定没有非null值吗

Comment (Gabriel39) [reply be/src/vec/exec/ves_http_scan_node.cpp:93]: 噢噢，got it，我理解是这样，这个地方传进来的data肯定不会是null，因为在上层函数scanner_scan遍历整个tuple的时候，就已经把null值过滤掉了(line 212 ~ line 219)，所以这个地方nullable一定是0

File: be/src/vec/exec/ves_http_scan_node.cpp [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+            if (_scan_finished.load()) {
+                return Status::OK();
+            }
+
+            // Get from scanner
+            RETURN_IF_ERROR(scanner->get_next(columns, tuple_pool.get(), &scanner_eof, _docvalue_context));
+        }
+
+        if (columns[0]->size() > 0) {
+            std::unique_lock<std::mutex> l(_block_queue_lock);
```
Comment (HappenLee) [inline be/src/vec/exec/ves_http_scan_node.cpp:118]: 这个锁锁太久了，  (*_vconjunct_ctx_ptr)->execute(block.get(), &result_column_id);
                Block::filter_block(block.get(), result_column_id, _tuple_desc->slots().size());
这两个方法不应该在锁里执行的

**Pull request merged**